echo_client.py

```diff
@@ -5,41 +5,30 @@
 
 def client(msg, log_buffer=sys.stderr):
     server_address = ('localhost', 10000)
-    # TODO: Replace the following line with your code which will instantiate
-    #       a TCP socket with IPv4 Addressing, call the socket you make 'sock'
-    sock = None
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
     print('connecting to {0} port {1}'.format(*server_address), file=log_buffer)
-    # TODO: connect your socket to the server here.
+    sock.connect(server_address)
 
-    # you can use this variable to accumulate the entire message received back
-    # from the server
     received_message = ''
 
-    # this try/finally block exists purely to allow us to close the socket
-    # when we are finished with it
     try:
         print('sending "{0}"'.format(msg), file=log_buffer)
-        # TODO: send your message to the server here.
-
-        # TODO: the server should be sending you back your message as a series
-        #       of 16-byte chunks. Accumulate the chunks you get to build the
-        #       entire reply from the server. Make sure that you have received
-        #       the entire message and then you can break the loop.
-        #
-        #       Log each chunk you receive.  Use the print statement below to
-        #       do it. This will help in debugging problems
-        chunk = ''
-        print('received "{0}"'.format(chunk.decode('utf8')), file=log_buffer)
+        sock.sendall(msg.encode('utf-8'))
+
+        while True:
+            chunk = sock.recv(16)
+            if not chunk:
```

Comment:
This approach will cause tests to hang, because _sock.recv()_ is a blocking call, meaning that instead of returning _None_ for an empty receive buffer (which is what you're checking for on line 20), it will sit there waiting for something to be received.

One alternative is to check the size of the receive buffer. If it is less than 16, it will indicate the tail end of a message, which you can use to trigger _break_. A corner case would be a message whose length is a multiple of 16, but we can live with that for the purposes of this assignment.